### PR TITLE
RL refit pipelining support

### PR DESCRIPTION
**Pull request by wdykas**
# What does this PR do ?
This enables changing the pipeline size between training and inference for RL. It also fixes a bug that the wrong model was being passes into the eval section when using refit causing it to be off.

Here are some verification runs that were done for comparing changing PP and TP to the baseline no refit

image

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @megatron-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [ef0e340](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/ef0e340a53f799478d2b5159ef26fbf45420a7fb)

**Comment by wdykas** [comment]
/ok to test [74d6cd2](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/74d6cd2c0132cced3f61dc3642a0389de15a0ed8)

**Comment by wdykas** [comment]
/ok to test [74d6cd2](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/74d6cd2c0132cced3f61dc3642a0389de15a0ed8)

**Comment by wdykas** [comment]
/ok to test [23dd22d](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/23dd22dce68a53f5d63f25b6fdafa8580d11e420)

File: megatron/core/inference/contexts/dynamic_context.py [hunk 1]
```diff
@@ -264,6 +265,7 @@ def __init__(
         block_size_tokens: int = 256,
         tensor_model_parallel_size: Optional[int] = None,
         pipeline_model_parallel_size: Optional[int] = None,
+        pg_collection: Optional[ProcessGroupCollection] = None,
```
Comment (wdykas) [inline megatron/core/inference/contexts/dynamic_context.py:268]: @santhnm2 For RL refit we are going to need to pass the pg_collection to grab the PP group, but this is starting to get a bit weird with like tensor_model_parallel_size, pipeline_model_parallel_size, and pg_collection all being passed. We could have this context only take in a pg_collection or we could just leave as is. Do you have any thoughts on how these should be handled?

**Comment by wdykas** [comment]
/ok to test [d825eab](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/d825eab1f4bb8284017787639d40912c9fcf3e49)

**Comment by wdykas** [comment]
/ok to test [d825eab](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/d825eab1f4bb8284017787639d40912c9fcf3e49)

**Comment by wdykas** [comment]
/ok to test [596e7db](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/596e7dbb7d52dd2749018f5682f79c835cfc998a)

File: megatron/core/inference/contexts/dynamic_context.py (hunk 1 above)
Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py:268]: I would be ok with just passing in the `pg_collection` and getting rid of the separate `tensor_model_parallel_size` and `pipeline_model_parallel_size` arguments. We would just have to be careful to update all the places where we initialize the context. Btw, I think the eventual long-term solution is moving to a model where we just pass the `TransformerConfig` (and/or an even finer-grained `InferenceConfig`) which would anyway remove the need for these individual arguments.

Comment (wdykas) [reply megatron/core/inference/contexts/dynamic_context.py:268]: ok we haven't unified everywhere we set this right so I need to go find it everywhere?
Also we still need to pass the pg_collection even if we pass TransformerConfig because this file uses the actual group it seems and not just the size

Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py:268]: Yea still need to unify all the calls :/ 
And yea I agree, we would still need to pass the `pg_collection`

Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py:268]: Maybe we can merge this in as is and then do a follow-up PR to clean up all the `DynamicInferenceContext` initializations + remove the TP / PP args?

Comment (wdykas) [reply megatron/core/inference/contexts/dynamic_context.py:268]: maybe let me see how big the blast radius is if I try fixing partially in the is MR. If its too big then I think its own clean up MR makes more sense or else this MR would grow to large.

**Comment by wdykas** [comment]
/ok to test [fbb89e4](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/fbb89e4a5c561cc7705fd27df35cbc08eb956ee5)

File: megatron/core/resharding/utils.py [hunk 2]
```diff
@@ -157,12 +175,81 @@ def assign_resolved_name_inplace(meta: ParameterMetadata) -> None:
...
+    layer module's `layer_number`.
+
+    Implementation:
+    - Cache a `{local_prefix -> global_prefix}` map on the root module.
+    - Perform a longest-prefix match replacement so we only rewrite the module path portion.
+    """
+    prefix_map = getattr(module, '_reshard_layer_module_prefix_map', None)
+    if prefix_map is None:
+        prefix_map = _build_layer_module_prefix_map(module)
+        setattr(module, '_reshard_layer_module_prefix_map', prefix_map)
```
Comment (tdene) [inline megatron/core/resharding/utils.py]: Instead of passing `module` in through the execution stack and running `_build_layer_module_prefix_map(module)` down here, I think it'd be a lot cleaner to grab this map up in `build_centralized_reshard_plan` where you already have the models and you can build the maps before you start looping over all the params.

Unless I'm misunderstanding, the `prefix_map` building looks to be param-agnostic, so it can be built up there?

File: megatron/rl/inference/megatron.py [hunk 3]
```diff
@@ -150,8 +154,8 @@ def get_dynamic_inference_engine(
         buffer_size_gb=args.inference_dynamic_batching_buffer_size_gb,
         max_requests=args.inference_dynamic_batching_max_requests,
         max_tokens=args.inference_dynamic_batching_max_tokens,
-        tensor_model_parallel_size=inference_tp_size,
-        pipeline_model_parallel_size=inference_pp_size,
+        # TP/PP sizes are derived from the model's pg_collection.
+        pg_collection=pg_collection,
```
Comment (tdene) [inline megatron/rl/inference/megatron.py]: complete nit: personally I strongly dislike full-line comments inside argument lists, and would rather see
```
pg_collection=pg_collection, # TP/PP sizes are derived from the model's pg_collection.
```

File: megatron/rl/parallel_utils.py [hunk 4]
```diff
@@ -29,11 +29,11 @@ def build_inference_pg_collection(
...
 
     Args:
-        tp_size: Tensor model parallel size for the inference model.
         world_size: Total world size (number of ranks).
+        tp_size: Tensor model parallel size for the inference model. Defaults to current MPU value if None.
+        pp_size: Pipeline parallel size. Defaults to current MPU value if None.
         use_tp_pp_dp_mapping: If True, use 'tp-cp-ep-pp-dp' order; otherwise 'tp-cp-ep-dp-pp'.
         cp_size: Context parallel size. Defaults to current MPU value if None.
-        pp_size: Pipeline parallel size. Defaults to current MPU value if None.
         ep_size: Expert parallel size. Defaults to current MPU value if None.
```
Comment (tdene) [inline megatron/rl/parallel_utils.py:36]: nit: reorder arg docstrings in the same order as the args themselves

File: megatron/rl/rl_utils.py [hunk 5]
```diff
@@ -1671,9 +1670,11 @@ def megatron_rl_inference_mode(
 
 
 def rl_inference_interface_shutdown():
+    global _INFERENCE_INTERFACE
     if _INFERENCE_INTERFACE is not None:
         loop = get_asyncio_loop()
         loop.run_until_complete(_INFERENCE_INTERFACE.kill())
+        _INFERENCE_INTERFACE = None
```
Comment (tdene) [inline megatron/rl/rl_utils.py:1677]: Agreed! Thank you!

File: megatron/training/training.py [hunk 6]
```diff
@@ -863,7 +874,9 @@ def pretrain(
         prefix = f'iteration {iteration} on validation set'
         if getattr(args, 'perform_rl_step', False):
             rl_utils.evaluate_and_print_results_rl(
-                valid_data_iterator, model, optimizer,
+                valid_data_iterator,
+                inference_model if inference_model is not None else model,
+                optimizer,
```
Comment (tdene) [inline megatron/training/training.py:889]: If we're changing this to use the `inference_model`, then we assume that we have already run `swap_model_weights`, correct?

I do not see how `swap_model_weights` could have been run at this point in time, right after we come out of `train()`.

File: megatron/training/arguments.py [hunk 7]
```diff
@@ -2115,6 +2115,12 @@ def _add_rl_args(parser):
                             'round-robin: distribute bins cyclically across ranks for better load balancing')
     group.add_argument('--rl-inference-tensor-model-parallel-size', type=int, default=None,
                        help='Degree of tensor model parallelism for inference for RL.')     
+    group.add_argument(
+        '--rl-inference-pipeline-model-parallel-size',
+        type=int,
+        default=None,
+        help='Degree of pipeline model parallelism for the separate RL inference model.',
```
Comment (yobibyte) [inline megatron/training/arguments.py]: Nit: replace 'separate'. From this descrption it's unclear, if we have another inference model for RL or something else.

File: megatron/core/resharding/utils.py [hunk 8]
```diff
@@ -157,12 +175,81 @@ def assign_resolved_name_inplace(meta: ParameterMetadata) -> None:
...
+    name = meta.name if base_name is None else base_name
+    if module is not None:
+        name = _resolve_global_layer_number_in_name(name, module)
+    assign_ep_resolved_name_inplace(meta, base_name=name)
+
+
+def _build_layer_module_prefix_map(module: torch.nn.Module) -> dict[str, str]:
+    """Build a mapping `{local_module_prefix -> global_module_prefix}` for PP layer modules.
+
+    Megatron assigns a *global*, 1-indexed `layer_number` to each transformer layer module at
```
Comment (yobibyte) [inline megatron/core/resharding/utils.py]: Not sure how stable this convention. Should we add a test/assert so that in the future when the 'global/1-indexed layer number' is changed by megatron we could quickly identify the issue?

E.g. if megatron switches to a zero-indexed numeration, we will get a negative number in here. We can check for that. 

But again, depends on how stable we believe the numeration convention is.

Comment (wdykas) [reply megatron/core/resharding/utils.py]: that indexing numeration will not change. It had been standard for a long time and would break a ton of things in Megatron if it was ever changed since its used in a ton of places

File: megatron/training/arguments.py (hunk 7 above)
Comment (wdykas) [reply megatron/training/arguments.py]: removed to match TP rl arg

File: megatron/training/training.py (hunk 6 above)
Comment (wdykas) [reply megatron/training/training.py:889]: yeah your right, I guess it looks ok in my runs since it would at most be off by a single optimizer step. I have added the swapping of weights properly in both places we do rl evaluation. Seems like I had missed the other call site as well so this was a good catch.

File: megatron/rl/parallel_utils.py (hunk 4 above)
Comment (wdykas) [reply megatron/rl/parallel_utils.py:36]: done

File: megatron/rl/inference/megatron.py (hunk 3 above)
Comment (wdykas) [reply megatron/rl/inference/megatron.py]: yup agreed. Done

File: megatron/core/resharding/utils.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/resharding/utils.py]: yeah like this. I think I did what you are referring to

**Comment by wdykas** [comment]
/ok to test [aa09290](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/aa092907f47e7f4faa788ed5986702e6aab30774)

File: megatron/core/resharding/utils.py (hunk 8 above)
Comment (yobibyte) [reply megatron/core/resharding/utils.py]: sounds good to me

**Comment by wdykas** [comment]
/ok to test [298ba7c](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/298ba7cd9d213007ed66aefa244894d884d813e2)

**Comment by wdykas** [comment]
/ok to test [486f88b](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/486f88bcdf191527fccee554156d4198abe83fcb)

**Comment by wdykas** [comment]
/ok to test [0eca251](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/0eca25152ac8b242b6e187b8ff3cc50c33890f0f)

**Comment by wdykas** [comment]
/ok to test [0eca251](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/0eca25152ac8b242b6e187b8ff3cc50c33890f0f)

**Comment by wdykas** [comment]
/ok to test [7a2fcdb](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/7a2fcdbc448951851af81a81b6c381cec632b53b)

**Comment by wdykas** [comment]
/ok to test [71fd639](https://github.com/NVIDIA/Megatron-LM/pull/2878/commits/71fd639db1d087262540d21525f7dcf5ac2ad09f)
